From b5d3839ed65a70de040ffa146d98fa3bf135b8e9 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:32 -0400
Subject: [fs] btrfs: fix wrong max system array size check in kernel space

Message-id: <1407179109-13714-326-git-send-email-sandeen@redhat.com>
Patchwork-id: 87322
O-Subject: [RHEL7.1 PATCH 325/422] btrfs: fix wrong max system array size check in kernel space
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Gui Hecheng <guihc.fnst@cn.fujitsu.com>

commit 5f43f86e3f62734dfeeb36c4d940425a2c6789dc
Author: Gui Hecheng <guihc.fnst@cn.fujitsu.com>
Date:   Mon Apr 21 20:13:11 2014 +0800

    btrfs: fix wrong max system array size check in kernel space

    For system chunk array,
    We copy a "disk_key" and an chunk item each time,
    so there should be enough space to hold both of them,
    not only the chunk item.

    Signed-off-by: Gui Hecheng <guihc.fnst@cn.fujitsu.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/volumes.c b/fs/btrfs/volumes.c
index dd07729..9ab1512 100644
--- a/fs/btrfs/volumes.c
+++ b/fs/btrfs/volumes.c
@@ -3921,7 +3921,8 @@ static int btrfs_add_system_chunk(struct btrfs_root *root,
  u8 *ptr;
 
  array_size = btrfs_super_sys_array_size(super_copy);
- if (array_size + item_size > BTRFS_SYSTEM_CHUNK_ARRAY_SIZE)
+ if (array_size + item_size + sizeof(disk_key)
+   > BTRFS_SYSTEM_CHUNK_ARRAY_SIZE)
   return -EFBIG;
 
  ptr = super_copy->sys_chunk_array + array_size;
-- 
1.7.1